### PR TITLE
Workaround Flux#1027

### PR DESCRIPTION
**Pull request by ericphanson**
This example very much does not work:
```julia
julia> output
10×1 Matrix{Float32}:
 1.2540959f-6
 5.1405354f-5
 0.00020041084
 3.131654f-5
 5.9973813f-6
 4.5418696f-7
 5.347429f-8
 0.99964094
 3.7541522f-5
 3.067953f-5

julia> output2
10×1 Matrix{Float32}:
 0.086956024
 0.10940457
 0.09150873
 0.098902285
 0.088835925
 0.06954686
 0.06539499
 0.20102207
 0.090939134
 0.0974893
```

The goal of this PR is to fix things so that we can correctly (de)-serialize this model so the test passes.

### CODE REVIEW

**Comment by ericphanson** [comment]
This works around https://github.com/FluxML/Flux.jl/issues/1027 by essentially following the strategy of https://github.com/FluxML/Flux.jl/pull/1509#issuecomment-798947090, where here I've called @touchesir's `actually_all_params` just `weights` and added `loadweights!` as an analog to `loadparams!` that uses these.

This adds a Flux dependency here which is a little unfortunate because downstream code might want to use the LegolasFlux schema without actually touching weights/flux stuff (e.g. looking at a bunch of losses or something), and now such code will pull in the Flux dependency. But when upstream fixes #1027 we can probably remove the `flux_workarounds.jl` file (and make a breaking release).

File: Project.toml [hunk 1]
```diff
@@ -1,22 +1,27 @@
 name = "LegolasFlux"
 uuid = "eb5f792d-d1b1-4535-bae3-d5649ec7daa4"
 authors = ["Beacon Biosignals, Inc."]
-version = "0.1.0"
+version = "0.1.1"
```
Comment (ericphanson) [inline Project.toml:4]: non-breaking; all existing usage of LegolasFlux will work, but if you switch to the new `weights` / `loadweights!` it'll even work correctly with `BatchNorm`s 😄

File: examples/digits.jl [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+
+
+# Increase to get more training/test data
+N_train = 1_000
+N_test = 50
+
+##
+# to use MNIST data, uncomment these
+# train_x, train_y = MNIST.traindata(Float32, 1:N_train)
+# test_x,  test_y  = MNIST.testdata(Float32, 1:N_test)
```
Comment (ericphanson) [inline examples/digits.jl:71]: decided to avoid downloading data in CI / any possible license issues with MNIST (MLDatasets gives you a warning that the license is ambiguous)

File: src/flux_workarounds.jl [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+
+"""
+    LegolasFlux.other_weights(layer) -> Vararg{Array}
+
+Given a layer with params that are not captured by `Flux.trainable`, produce
+a tuple of arrays corresponding to these parameters (analogous to `Flux.trainable`).
+"""
+function other_weights end
+
+other_weights(layer) = ()
```
Comment (ToucheSir) [inline src/flux_workarounds.jl:13]: Re Flux dependency, the only dependency that is strictly necessary here is Functors.jl itself. `Functors.children(layer)` will return _everything_ that layer considers a parameter, so you shouldn't be missing out on anything :)

File: src/flux_workarounds.jl [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+# The following is a copy of <https://github.com/FluxML/Flux.jl/blob/335286adf118b61ad6fffa5937bd9358477a00c9/src/functor.jl#L41-L63>
+# with `params` changed to `weights` and the addition of the lines
+# ```julia
+# for child in other_weights(x)
+#     weights!(p, child, seen)
+# end
+# ```
+# to `weights!(p::Params, x, seen = IdSet())`.
+
+weights!(p::Params, x::AbstractArray{<:Number}, seen = IdSet()) = push!(p, x)
```
Comment (ToucheSir) [inline src/flux_workarounds.jl:31]: Instead of pushing to `Params` (which requires a dep on Zygote), you can just push to a plain vector of arrays. I imagine that would also be much easier to serialize as well!

**Review by ToucheSir — COMMENTED** [review]
Thanks for the ping! I've left a couple of comments that might be helpful.

File: src/flux_workarounds.jl [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+using Base: IdSet
+export weights, loadweights!
+
+"""
+    LegolasFlux.other_weights(layer) -> Vararg{Array}
+
+Given a layer with params that are not captured by `Flux.trainable`, produce
+a tuple of arrays corresponding to these parameters (analogous to `Flux.trainable`).
+"""
+function other_weights end
```
Comment (hannahilea) [inline src/flux_workarounds.jl:11]: what does this line do, given the default implementation below?

File: src/flux_workarounds.jl (hunk 4 above)
Comment (ToucheSir) [reply src/flux_workarounds.jl:31]: I just occurred to me that `weights!` is very similar to [`fcollect(...; exclude=Functors.isleaf)`](https://github.com/FluxML/Functors.jl/blob/master/src/functor.jl#L108). Would that work for your use-case?

File: test/runtests.jl [hunk 6]
```diff
@@ -43,3 +45,37 @@ end
...
+        testmode!(r1)
+
+        @test output ≈ r1(x)
+
+        if layer == BatchNorm
+            r2 = mk_model()
+            Flux.loadparams!(r2, p)
+            testmode!(r2)
+
+            @test_broken output ≈ r2(x)
```
Comment (hannahilea) [inline test/runtests.jl:74]: Nice! Might be worth adding a comment here that iff this test _fails_, we can reevaluate use of flux workaround, and linking to the description you wrote in this issue?

**Review by hannahilea — CHANGES_REQUESTED** [review]
Nice!

Other than the renaming issue below, this seems good to go. :)

File: src/flux_workarounds.jl (hunk 5 above)
Comment (ericphanson) [reply src/flux_workarounds.jl:11]: I did this so the docstring would be attached to the function as a whole, instead of a particular method. It doesn't matter here since none of the others have docstrings anyway, but we could add a separate docstring for each method and if you do `? other_weights` or `? other_weights(layer)` it will show you all the docstrings for all the methods that fit that signature, so we could just put it on the fallback method and it would always show up, just like it will when we attach it to the function itself like I did here (see https://docs.julialang.org/en/v1/manual/documentation/#Functions-and-Methods for probably a better explanation). But really I just did it because I felt like conceptually that docstring belongs to the whole function and I felt like it helped me formulate the API. Like here's the function, here's what it does in words, and then here are the methods to do that.

File: test/runtests.jl (hunk 6 above)
Comment (ericphanson) [reply test/runtests.jl:79]: done!

**Comment by ericphanson** [comment]
Thanks for the reviews!

Thanks especially for the pointers to `fcollect` and Functors, @ToucheSir! Super helpful. In the last two commits, I switched to using that approach and dropped the Flux dep (for the very light Functors dep). I found that `fcollect` as-is does not work because of this line: https://github.com/FluxML/Functors.jl/blob/adeb24bc3b2fb3e9959f1157d81f4633a855e207/src/functor.jl#L109. `x in arr` for arrays `arr` compares by value, and if we end up with several `x`s with the same values (say zero'd out arrays from a freshly initialized model or something like that), then we only keep 1 of each, and when it comes time to load the weights later, we don't have enough of them (this isn't hypothetical, it happened with the DigitsModel example here and had me confused for awhile). So instead in [`fcollect2`](https://github.com/beacon-biosignals/LegolasFlux.jl/pull/4/files#diff-1762840cbf23144855a52c56e365f6fc4358ea3d6307893157c2e6027aa73953R1-R11) I switched to using an `IdSet` to keep track of the objects we've already added (similarly to `fmap` above). Since an `IdSet` is unordered and we definitely care about the ordering, I had to also keep the original `Vector{Any}` used to collect the weights. Keeping them in two data structures is a little inelegant (I guess I actually want an `OrderedIdSet` which doesn't exist AFAIK) but should have a negligible perfomance cost in this context.

Also, since we now have `loadweights!` available in the package, we might be able to make the API a little easier to use (e.g. allow a user to pass a model directly instead of making them get the weights out first themselves). But I think that should be followup work (I think it requires a little more thought about the best approach).

**Comment by ToucheSir** [comment]
Nice catch for `fcollect`. IMO this also a bug in the upstream implementation and we should fix it to work like `fcollect2`. Will have a look at that when I next find time to work on Functors.

**Comment by ericphanson** [comment]
Ok good to hear this is likely a bug! I've filed an issue, https://github.com/FluxML/Functors.jl/issues/16.

File: README.md [hunk 7]
```diff
@@ -28,16 +29,16 @@ my_model = make_my_model()
 using LegolasFlux
 
 # We can save whatever other columns we'd like to as well as the `weights`.
-model_row = ModelRow(; weights = collect(params(cpu(my_model))), architecture_version = 1, loss = 0.5)
+model_row = ModelRow(; weights = weights(cpu(my_model)),
+                       architecture_version = 1, loss = 0.5)
```
Comment (ararslan) [inline README.md:33]:
```suggestion
                     architecture_version=1, loss=0.5)
```

File: examples/digits.jl [hunk 8]
```diff
@@ -0,0 +1,140 @@
...
+    config::DigitsConfig
+end
+
+# Ensure Flux can recurse into our model to find params etc
+Flux.@functor DigitsModel (chain,)
+
+# Construct the actual model from a config object. This is the only
+# constructor that should be used, to ensure the model is created just
+# from the config object alone.
+function DigitsModel(config::DigitsConfig = DigitsConfig())
```
Comment (ararslan) [inline examples/digits.jl:33]:
```suggestion
function DigitsModel(config::DigitsConfig=DigitsConfig())
```

File: examples/digits.jl [hunk 9]
```diff
@@ -0,0 +1,140 @@
...
+        Dropout(dropout_rate),
+        Conv((3, 3), 32=>16, relu),
+        Dropout(dropout_rate),
+        MaxPool((2,2)),
+        Dropout(dropout_rate),
+        Conv((3, 3), 16=>10, relu),
+        Dropout(dropout_rate),
+        x -> reshape(x, :, size(x, 4)),
+        Dropout(dropout_rate),
+        Dense(90, 10), softmax)
```
Comment (ararslan) [inline examples/digits.jl:50]:
```suggestion
    chain = Chain(Dropout(dropout_rate),
                  Conv((3, 3), 1 => 32, relu),
                  BatchNorm(32, relu),
                  MaxPool((2, 2)),
                  Dropout(dropout_rate),
                  Conv((3, 3), 32 => 16, relu),
                  Dropout(dropout_rate),
                  MaxPool((2, 2)),
                  Dropout(dropout_rate),
                  Conv((3, 3), 16 => 10, relu),
                  Dropout(dropout_rate),
                  x -> reshape(x, :, size(x, 4)),
                  Dropout(dropout_rate),
                  Dense(90, 10),
                  softmax)
```

File: examples/digits.jl [hunk 10]
```diff
@@ -0,0 +1,140 @@
...
+    val = mean(onecold(m(x)) .== onecold(y))
+    trainmode!(m)
+    return val
+end
+
+function train_model!(m; N = N_train)
+    loss = (x, y) -> crossentropy(m(x), y)
+    opt = ADAM()
+    evalcb = throttle(() -> @show(accuracy(m, tX, tY)), 5)
+    Flux.@epochs 1 Flux.train!(loss, params(m), Iterators.take(train, N), opt, cb = evalcb)
```
Comment (ararslan) [inline examples/digits.jl:116]:
```suggestion
    Flux.@epochs 1 Flux.train!(loss, params(m), Iterators.take(train, N), opt; cb=evalcb)
```

File: examples/digits.jl [hunk 11]
```diff
@@ -0,0 +1,140 @@
...
+    return accuracy(m, tX, tY)
+end
+
+m = DigitsModel()
+
+# increase N to actually train more than a tiny amount
+acc = train_model!(m; N = 10)
```
Comment (ararslan) [inline examples/digits.jl:123]:
```suggestion
acc = train_model!(m; N=10)
```

File: src/functors.jl [hunk 12]
```diff
@@ -0,0 +1,39 @@
+# Modified version of `fcollect` to use an `IdSet` cache so that
+# distinct arrays whose values happen to be duplicates are each kept.
+# <https://github.com/FluxML/Functors.jl/issues/16>
+function fcollect2(x; output = [], cache = Base.IdSet(), exclude = v -> false)
```
Comment (ararslan) [inline src/functors.jl:4]:
```suggestion
function fcollect2(x; output=[], cache=IdSet(), exclude=_ -> false)
```

File: src/functors.jl [hunk 13]
```diff
@@ -0,0 +1,39 @@
...
+    x in cache && return output
+    if !exclude(x)
+      push!(cache, x)
+      push!(output, x)
+      foreach(y -> fcollect2(y; cache = cache, output=output, exclude = exclude), Functors.children(x))
```
Comment (ararslan) [inline src/functors.jl:9]:
```suggestion
        push!(cache, x)
        push!(output, x)
        foreach(y -> fcollect2(y; cache=cache, output=output, exclude=exclude), Functors.children(x))
```

File: src/functors.jl [hunk 14]
```diff
@@ -0,0 +1,39 @@
...
+
+"""
+    loadweights!(m, xs)
+
+Load weights `xs` into the model `m`, using [`weights`](@ref).
+"""
+function loadweights!(m, xs)
+    for (i, (p, x)) in enumerate(zip(weights(m), xs))
+        if size(p) != size(x)
+            error("For the $(i)th weight expected param size $(size(p)), got $(size(x))")
```
Comment (ararslan) [inline src/functors.jl:32]: Perhaps throw an informative error type, e.g. `ArgumentError`?

File: test/runtests.jl [hunk 15]
```diff
@@ -43,3 +46,42 @@ end
...
     @test Arrow.Table(Arrow.tobuffer(tbl)).weights[1] == w
 end
+
+@testset "`flux_workarounds`" begin
+    @testset "layer $layer" for layer in [BatchNorm, InstanceNorm, (c) -> GroupNorm(c, 1), c -> identity]
+        mk_model = () -> (Random.seed!(1); Chain(Dense(1, 10), Dense(10, 10), layer(1), Dense(10, 1)))
+        model = mk_model()
+        trainmode!(model)
+        x = reshape([1f0], 1, 1, 1)
+        for i = 1:10
```
Comment (ararslan) [inline test/runtests.jl:56]:
```suggestion
        for i in 1:10
```

File: src/functors.jl [hunk 16]
```diff
@@ -0,0 +1,43 @@
...
+
+"""
+    load_weights!(m, xs)
+
+Load weights `xs` into the model `m`, using [`weights`](@ref).
+"""
+function load_weights!(m, xs)
+    model_weights = weights(m)
+    if length(model_weights) != length(xs)
+        throw(ArgumentError("Number of weights given ($(length(xs))) does not match number of weights model expects ($(length(model_weights)))"))
```
Comment (hannahilea) [inline src/functors.jl:32]: :)

File: test/runtests.jl [hunk 17]
```diff
@@ -3,6 +3,7 @@ using Test
 using Flux, LegolasFlux
 using LegolasFlux: Weights, FlatArray, ModelRow
 using Arrow
+using Random
```
Comment (hannahilea) [inline test/runtests.jl:6]: Is it possible we want to use StableRNG here (and then seeded rands, below)?

File: src/flux_workarounds.jl (hunk 5 above)
Comment (hannahilea) [reply src/flux_workarounds.jl]: oh cool! thanks for explaining. :)

File: src/functors.jl (hunk 14 above)
Comment (ericphanson) [reply src/functors.jl]: done!

File: test/runtests.jl (hunk 17 above)
Comment (ericphanson) [reply test/runtests.jl:6]: good idea, done!

**Review by hannahilea — CHANGES_REQUESTED** [review]
...so I wrote out a whole big comment and was _sure_ it was attached to my previous review, but now I don't see it anywhere? The gist is that `weights` has a very specific meaning in the context of ML, so our use of `weights` (in `load_weights(...)`, `weights(...)`, etc) should probably actually be renamed to something more like `parameters` or `values`. (Too bad `params` is already taken by Flux!)

Basically, while all weights are parameters, not all parameters are weights (some are biases, etc).

File: src/functors.jl (hunk 16 above)
Comment (hannahilea) [reply src/functors.jl:32]: safety first! :)

**Comment by ToucheSir** [comment]
> (Too bad `params` is already taken by Flux!)

Not only that, the name `params` is misleading because it only returns trainable parameters. Moreover, it conflicts with `BenchmarkTools.params` as well. I wish we could rename it to something else, but it would be a massive back compat break because `params` is arguably the most used function in all of Flux (not counting re-exports from Zygote etc.)

File: src/functors.jl (hunk 16 above)
Comment (ericphanson) [reply src/functors.jl:32]: 💯

Comment (ToucheSir) [reply src/functors.jl:32]: Yes please, not doing this has lead to subtle errors in `loadparams!` in upstream Flux :)

**Comment by ericphanson** [comment]
Ok, we've got some bikeshedding to do! Some ideas...

|           | Getter            | Setter            | Column name |
|-----------|-------------------|-------------------|-------------|
| Current   | `weights`         | `load_weights!`   | `weights`   |
| state     | `fetch_state`     | `load_state!`     | `state`     |
| learnings | `fetch_learnings` | `load_learnings!` | `learnings` |

We're leaning towards `state` but if anyone has any further ideas or comments let me know! I will pick something and merge tomorrow morning if there aren't other objections.

**Comment by hannahilea** [comment]
...pytorch uses `state`:
> In PyTorch, the learnable parameters (i.e. weights and biases) of an torch.nn.Module model are contained in the model’s parameters (accessed with model.parameters()). A state_dict is simply a Python dictionary object that maps each layer to its parameter tensor. Note that only layers with learnable parameters (convolutional layers, linear layers, etc.) and registered buffers (batchnorm’s running_mean) have entries in the model’s state_dict. Optimizer objects (torch.optim) also have a state_dict, which contains information about the optimizer’s state, as well as the hyperparameters used.
 (https://pytorch.org/tutorials/beginner/saving_loading_models.html#what-is-a-state-dict)

...tensorflow uses `weights` (huh):
> A Keras model consists of multiple components:
> 
> The architecture, or configuration, which specifies what layers the model contain, and how they're connected.
> A set of weights values (the "state of the model"). (https://www.tensorflow.org/guide/keras/save_and_serialize#introduction)

so maybe I sent us down this bikeshed prematurely, and it is okay to stick with `weights` after all... 🤷

**Review by hannahilea — APPROVED** [review]
...in light of the renaming discussion above, I'm going to leave the renaming (or not) to your discretion and let you merge at your leisure.

**Comment by ericphanson** [comment]
Ok cool! Thanks for looking up how others do it, that was a great idea.

I think then since `weights` seems acceptable I'd like to go with that. We only allow (multi-dimensional) arrays with all the same element type, which is far from arbitrary state. That restriction is on purpose of course, since we are trying to provide a different way to serialize the model rather than include arbitrary state from your julia session, so I think it's good to highlight it a little. Also, it's nice to keep this non-breaking by not adjusting the schema.

But I think I'd like to use `fetch_weights` instead of `weights` so that we have `fetch_weights`, `load_weights!`, and column name `weights`. I like that better than reusing `weights` for both the function and the column name because you can then do
```julia
weights = fetch_weights(model)

row = ModelRow(; weights, ...)
```
